From fc39a29b6d1748aaedf7d5358848aa51f95d5bea Mon Sep 17 00:00:00 2001
From: Fabian Hartung <fabian.hartung@chargebyte.com>
Date: Thu, 13 Mar 2025 15:48:44 +0100
Subject: [PATCH] Add PRE_CHARGE_STARTED signal

This signal is necessary to inform that the pre-charge
process has started.

Signed-off-by: Fabian Hartung <fabian.hartung@chargebyte.com>
---
 include/iso15118/d20/state/dc_pre_charge.hpp | 1 +
 include/iso15118/session/feedback.hpp        | 1 +
 src/iso15118/d20/state/dc_pre_charge.cpp     | 5 +++++
 3 files changed, 7 insertions(+)

diff --git a/include/iso15118/d20/state/dc_pre_charge.hpp b/include/iso15118/d20/state/dc_pre_charge.hpp
index 38d3128..d8f6e6f 100644
--- a/include/iso15118/d20/state/dc_pre_charge.hpp
+++ b/include/iso15118/d20/state/dc_pre_charge.hpp
@@ -14,6 +14,7 @@ struct DC_PreCharge : public FsmSimpleState {
     HandleEventReturnType handle_event(AllocatorType&, FsmEvent) final;
 
 private:
+    bool pre_charge_initiated{false};
     float present_voltage{0};
 };
 
diff --git a/include/iso15118/session/feedback.hpp b/include/iso15118/session/feedback.hpp
index 63a2ca2..b51edb3 100644
--- a/include/iso15118/session/feedback.hpp
+++ b/include/iso15118/session/feedback.hpp
@@ -19,6 +19,7 @@ enum class Signal {
     REQUIRE_AUTH_EIM,
     START_CABLE_CHECK,
     SETUP_FINISHED,
+    PRE_CHARGE_STARTED,
     CHARGE_LOOP_STARTED,
     CHARGE_LOOP_FINISHED,
     DC_OPEN_CONTACTOR,
diff --git a/src/iso15118/d20/state/dc_pre_charge.cpp b/src/iso15118/d20/state/dc_pre_charge.cpp
index 8a75842..115c756 100644
--- a/src/iso15118/d20/state/dc_pre_charge.cpp
+++ b/src/iso15118/d20/state/dc_pre_charge.cpp
@@ -51,6 +51,11 @@ FsmSimpleState::HandleEventReturnType DC_PreCharge::handle_event(AllocatorType&
     const auto variant = ctx.pull_request();
 
     if (const auto req = variant->get_if<message_20::DC_PreChargeRequest>()) {
+        if (not pre_charge_initiated) {
+            ctx.feedback.signal(session::feedback::Signal::PRE_CHARGE_STARTED);
+            pre_charge_initiated = true;
+        }
+
         const auto res = handle_request(*req, ctx.session, present_voltage);
 
         ctx.feedback.dc_pre_charge_target_voltage(message_20::datatypes::from_RationalNumber(req->target_voltage));
-- 
2.34.1

